rpcs3/Emu/Cell/lv2/sys_net/lv2_socket_p2ps.cpp

```diff
@@ -0,0 +1,800 @@
+							auto& sock_p2ps = reinterpret_cast<lv2_socket_p2ps&>(sock);
+							sock_p2ps.set_status(p2ps_stream_status::stream_closed); });
+					it = msgs.erase(it);
+					continue;
+				}
+
+				// resend the message
+				const auto res = idm::check<lv2_socket>(msg.sock_id, [&](lv2_socket& sock) -> bool
+					{
+						ensure(sock.get_type() == SYS_NET_SOCK_STREAM_P2P);
+						auto& sock_p2ps = reinterpret_cast<lv2_socket_p2ps&>(sock);
+
+						if (sendto(sock_p2ps.get_socket(), reinterpret_cast<char*>(msg.data.data()), msg.data.size(), 0, reinterpret_cast<const sockaddr*>(&msg.dst_addr), sizeof(msg.dst_addr)) == -1)
+						{
+							sock_p2ps.set_status(p2ps_stream_status::stream_closed);
+							return false;
+						}
+						return true; });
+
+				if (!res || !res.ret)
+				{
+					it = msgs.erase(it);
+					continue;
+				}
+
+				// Update key timeout
+				msgs.insert(std::make_pair(now + rtt.rtt_time, std::move(msg)));
+				it = msgs.erase(it);
+			}
+		}
+	}
+
+	tcp_timeout_monitor& operator=(thread_state)
+	{
+		wakey.notify_one();
+		return *this;
+	}
+
+public:
+	static constexpr auto thread_name = "Tcp Over Udp Timeout Manager Thread"sv;
+
+private:
+	std::condition_variable wakey;
+	std::mutex data_mutex;
+	// List of outgoing messages
+	struct message
+	{
+		s32 sock_id;
+		::sockaddr_in dst_addr;
+		std::vector<u8> data;
+		u64 seq;
+		steady_clock::time_point initial_sendtime;
+	};
+	std::map<steady_clock::time_point, message> msgs; // (wakeup time, msg)
+	// List of rtts
+	struct rtt_info
+	{
+		unsigned long num_retries          = 0;
+		std::chrono::milliseconds rtt_time = 50ms;
+	};
+	std::unordered_map<s32, rtt_info> rtts; // (sock_id, rtt)
+};
+
+void initialize_tcp_timeout_monitor()
+{
+	g_fxo->need<named_thread<tcp_timeout_monitor>>();
+}
+
+u16 u2s_tcp_checksum(const u16* buffer, usz size)
+{
+	u32 cksum = 0;
+	while (size > 1)
+	{
+		cksum += *buffer++;
+		size -= sizeof(u16);
+	}
+	if (size)
+		cksum += *reinterpret_cast<const u8*>(buffer);
+
+	cksum = (cksum >> 16) + (cksum & 0xffff);
+	cksum += (cksum >> 16);
+	return static_cast<u16>(~cksum);
+}
+
+std::vector<u8> generate_u2s_packet(const p2ps_encapsulated_tcp& header, const u8* data, const u32 datasize)
+{
+	const u32 packet_size = (sizeof(u16) + sizeof(p2ps_encapsulated_tcp) + datasize);
+	ensure(packet_size < 65535); // packet size shouldn't be bigger than possible UDP payload
+	std::vector<u8> packet(packet_size);
+	u8* packet_data = packet.data();
+
+	*reinterpret_cast<le_t<u16>*>(packet_data) = header.dst_port;
+	memcpy(packet_data + sizeof(u16), &header, sizeof(p2ps_encapsulated_tcp));
+	if (datasize)
+		memcpy(packet_data + sizeof(u16) + sizeof(p2ps_encapsulated_tcp), data, datasize);
+
+	auto* hdr_ptr     = reinterpret_cast<p2ps_encapsulated_tcp*>(packet_data + sizeof(u16));
+	hdr_ptr->checksum = 0;
+	hdr_ptr->checksum = u2s_tcp_checksum(utils::bless<u16>(hdr_ptr), sizeof(p2ps_encapsulated_tcp) + datasize);
```

Comment:
why two assignments?

Reply:
Because hdr_ptr->checksum is used in the calculation of the checksum.